### ISSUE TITLE
zlib: implement partial flush

### ISSUE THREAD

**Issue by hanwen** [golang/go#31514]
### What version of Go are you using (`go version`)?

go-1.11.5

while trying to implement zlib compression for SSH (by internal Google request, 
https://github.com/golang/go/issues/31369), I stumbled on lack of support for partial flush.

partial flush is described more precisely here: https://www.bolet.org/~pornin/deflate-flush.html (look for Z_PARTIAL_FLUSH). While this is deprecated functionality, it has also made its way into the SSH RFC.

For the write side, it looks like I can't truncate the sync marker from a full flush to get the same effect.

For the read side, feeding packets from OpenSSH into zlib.Reader leads to errors ("unexpected EOF").

**Comment by bradfitz** [comment]
I assume this is in the flate package where you need this? At least initially?

/cc @nigeltao

**Comment by hanwen** [comment]
From what I could tell, the real work is in the flate package, but then the zlib package should also also expose it, because SSH produces and consumes the zlib version header.

**Comment by dsnet** [comment]
What about the reader side of things? Wouldn't a partial flush on the writer only be useful if the reader had API to synchronize on these specialized flushing modes?

I should note that this is more difficult on the reader side for API design, since the reader lacks an exported concrete type.

**Comment by dsnet** [comment]
Also, my reading of partial flush suggests that partial flush is a subset of sync flush, which the writer already supports.

**Comment by hanwen** [comment]
I've seen that the bytes already get decompressed correctly (so it kind of works), but then the reader wants to read on and finds EOF. The reader should know to stop further reads when it encounters a partial flush. 

For API: when the reader encounters partial flush, it could just return the bytes thus far read without error. In SSH then, the caller could use bytes.Buffer{} as input to the reader, and stop requesting further zlib bytes when the buffer is empty.

**Comment by dsnet** [comment]
@hanwen, is there internal framing within the SSH protocol such that it can distinguish between a return from `Read` due to a partial flush, as a opposed a return due to other reasons (i.e., full buffer)?

**Comment by hanwen** [comment]
SSH protocol has packets, that are up to 32kb in size. The entire packet payload is compressed, so one would read until the entire packet was processed. 

I think there is could be chunking going on, so that you may need to do multiple reads to process a single packet.

**Comment by dsnet** [comment]
> The entire packet payload is compressed, so one would read until the entire packet was processed.

I'm not sure if that answered my question. More specifically: how does SSH handle framing for the packet? For the un-compressed data does it have a length-prefix or a delimiter marking the length or end of the packet? Or does it rely on a partial flush in DEFLATE to determine the end of a packet?

**Comment by hanwen** [comment]
there is no size prefix or delimiter. On the read side, you just keep reading until there are no compressed bytes left to process. If the read API would be changed to be 

  // Read up to buf bytes, or up to the next partial or full flush encountered in the compressed input 
  // stream
  Read(buf []byte)

then you could do 

```
  // init
  buf := bytes.NewBuffer(compressed)
  reader := zlib.Reader(buf)
 ..


// handle packet:
  buf.Write(compressedPacket)
  var uncompressed []byte
  for buf.Len() > 0 {
      var out [10240]byte 
      n, err := reader.Read(out[:])
      // err handling
      uncompressed = append(uncompressed, out[:n]...)
  }
```

this might break current users that expect to be able to read a certain amount of data in a single Read call, though.

There is packet framing, but it is around the compressed payload: the compressed data is padded, encrypted, mac'd and prefixed with total length for the encrypted packet.

**Comment by branelmoro** [comment]
@hanwen if we put limit on number of bytes read from underlying Reader and return error if maximum read limit is reached, then I think this will solve problem of reading till EOF.

We will need new interface for decompressor:

```
type Limiter interface {
    // Sets limit on number of bytes read from underlying Reader
    SetReadLimit(n int) error
}
```

Also, we need to add new property to store limit on number of bytes read from underlying Reader.

```
type decompressor struct {
    ...
    l int
    ...
    ...
}
```

if `l < 0` (default -1), means that no read limit is set
if `l == 0`, means that maximum read limit reached on underlying Reader
if `l > 0`, means that maximum `l` number of bytes can be read from underlying Reader and it will return ReadLimitReached error if `l` number of bytes and already read and further read is done

Add `f.l = -1` as default value in below two functions:

```
func NewReader(r io.Reader) io.ReadCloser {
    ...
    f.l = -1
    ...
    ...
}

func NewReaderDict(r io.Reader, dict []byte) io.ReadCloser {
    ...
    f.l = -1
    ...
    ...
}
```

We will need two more functions for decompressor:

```
func (f *decompressor) SetReadLimit(n int) {
    f.l = n
}

func (f *decompressor) readByte() (byte, error) {
    if f.l == 0 {
        return 0, errors.New("ReadLimitReached")
    }
    b, e := f.r.ReadByte()
    if e == nil && f.l != -1{
        f.l -= 1
    }
    return b, e
}
```

And finally, we need to replace `f.r.ReadByte()` with `f.readByte()`





Then in application, you can do

```
import "compress/flate"

decompressor := flate.NewReader(compressed_buffer)
size := 10 // or size := compressed_buffer.Len() or some known size of partial flush of compressed_buffer
decompressor.(flate.Limiter).SetReadLimit(size)
_, err = io.Copy(decompressed_buffer, decompressor)
```

Above code will return ReadLimitReached error after reading 10 bytes but it will not read till EOF
Now you can use decompressor further by setting limit again or by resetting underlying Reader

**Comment by branelmoro** [comment]
@hanwen I have created pull request for this - https://github.com/golang/go/pull/33195
I hope this works...

**Comment by nhooyr** [comment]
The docs on the synchronization example at https://golang.org/pkg/compress/flate/ need to be updated.

They state:

```go
			// Read the message length.
			// This is guaranteed to return for every corresponding
			// Flush and Close on the transmitter side.
```

But that is not true given for a Flush where no bytes have been written.

See https://github.com/golang/go/blob/4b4370066f97ac40a4f095ca56a7b21a375aa5aa/src/compress/flate/inflate.go#L336

The loop only returns if there are additional bytes in the buffer to be read.

**Comment by nhooyr** [comment]
In order to implement context takeover compression for WebSockets, I was able to get around the above issue by using `panic` in the underlying `io.Reader` after the sync marker is read and then recovering. The reader remains useable after this with the correct sliding window.

Janky but works for now.

**Comment by nigeltao** [comment]
I think it's @dsnet's call about what to do re @nhooyr's latest comments.

**Comment by diamondburned** [comment]
Funnily enough, the hack of ignoring `io.ErrUnexpectedEOF` would've worked very well if `(compress/flate).Reader` doesn't check the last error on later reads.

**Comment by hhrutter** [comment]
There are also PDF files out there with seemingly embedded partial flushed flate encoded streams.
Hard to tell the source for these because of various PDFWriters out there. 
What's the current state of this issue?

**Comment by mjl-** [comment]
I added a NewReaderPartial method at https://github.com/mjl-/flate. It returns data written in "partial flush" mode to the reader without waiting for a "sync flush" (zero-length data) block or EOF or for the history/sliding window to be full. NewReaderPartial processes all flate blocks it sees in its buffered compressed input, but won't attempt to read another flate block (which may block indefinitely) if the input buffer is empty and it has decompressed data it can return.

The package does not have a writer that can do "partial flushes". It isn't essential: The other side will handle sync flushes just fine if it can handle partial flushes. But a partial flushing would still be nice to have, as it uses fewer bytes for protocols without outside framing (the sync flush emits at least 4 bytes for its zero-length data block). Based on explanations I read, writing partial flushes compatibly with existing implementations requires careful handling.

I wonder if/doubt that anyone really depends on "Read doesn't return before a sync flush (but only as long as the read is smaller than the history size)". Returning data from partial blocks may have been a reasonable default, but I can understand hesitation in changing existing behaviour.

The context: I need incoming data early, for compressed IMAP connections. I am writing an IMAP server (mox) that implements the "compress deflate" command, which enables deflate on the entire connection. IMAP is line/text oriented. Clients write a compressed line with a command, there is no other framing. It turns out that some clients flush in "partial mode", others in "sync mode". During testing, with the Go stdlib flate reader, the clients doing a "sync flush" after writing a command worked fine, but those doing "partial flush" would hang after writing their command. And that's because mox never got the command from the flate reader, because that was waiting for a sync flush. The data would come in, along with the unexpected eof error only when the client assumed the connection had timed out and the client closed it (unblocking flate's read for the next flate block).

If there is interest, I can make a CL for adding NewReaderPartial or a similar approach to the stdlib flate package.

@hhrutter Just curious why need partial flush support. With PDF files, don't you have the full contents of the deflate stream, including the end? I would expect that end of stream returns all remaining unread data to you. Unless the ending is missing...
